### PR TITLE
Use zmq

### PR DESCRIPTION
**Pull request by oswynb**
Rewrite of ceilometer-publisher-rabbitmq to instead use ZeroMQ (and also rename to ceilometer-publisher-zeromq). Acts as a client (ZMQ.Req type socket) and is designed to interact with vaultaire-collector-ceilometer, (which uses a ZMQ.Rep type socket).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,21 +11,19 @@ The traditional Ceilometer publisher, which sends data to MySQL or MongoDB did
...
 1. Openstack produces a JSON message blob
-2. Enters the Ceilometer pipeline either from a Telemetry (Ceilometer) or
-   Compute (Nova?) node
-3. Is published to the `rabbitmq` sink
-4. JSON message blob is enqueued to RabbitMQ (this is also durable; desirable)
-5. Reader process reads from the RabbitMQ exchange, performs any necessary
-   processing, then writes the output to a spoolfile via libmarquise
-6. Spoolfiles are read and shipped to Vaultaire by a marquised process.
+2. Enters the Ceilometer pipeline via a ceilometer agent
+3. ceilometer-publisher-zeromq sends the JSON to a collector
```
Comment (thsutton) [inline README.md]: "...over a ZeroMQ connection"

**Comment by thsutton** [comment]
Looks good to me.

File: ceilometer_publisher_zeromq/queue.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from ceilometer.openstack.common import log
+from oslo.config import cfg
+import json
+import zmq
+
+LOG = log.getLogger(__name__)
+
+OPTS = [
+    cfg.StrOpt('publisher_zeromq_host',
+               default='127.0.0.1'),
```
Comment (fractalcat) [inline ceilometer_publisher_zeromq/queue.py]: Perhaps 'localhost', so we're not assuming IPv4?

File: ceilometer_publisher_zeromq/queue.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, parsed_url):
+        LOG.info("Queue Publisher starting up")
+        super(QueuePublisher, self).__init__(parsed_url)
+        self.context = None
+        self.socket  = None
+        self.reconnect()
+
+    def reconnect(self):
+        """(re)connects to the collector"""
+        self.context = zmq.Context()
```
Comment (fractalcat) [inline ceilometer_publisher_zeromq/queue.py]: Do we want to shut down the old context before creating a new one? Not sure when the old one would get garbage-collected, though iirc it's safe to have multiple zmq contexts.

Comment (tranma) [inline ceilometer_publisher_zeromq/queue.py]: it seems to be quite expensive to create/destroy a ZMQ context, so we should use the one context for multiple sockets, as long it's single-threaded.

File: ceilometer_publisher_zeromq/queue.py (hunk 2 above)
Comment (oswynb) [inline ceilometer_publisher_zeromq/queue.py]: I don't think we can. ZMQ doesn't support the use of localhost, addresses are tied to transport and for tcp they much be numeric.

Comment (christian-marie) [inline ceilometer_publisher_zeromq/queue.py]: It supports the use of domain name lookups, perhaps localhost is not configured in /etc/hosts is the problem?

Comment (christian-marie) [inline ceilometer_publisher_zeromq/queue.py]: Either way this is not a big deal. We're just going to strace this if we get confused and are using IPV6, this is not a show-stopper.

File: ceilometer_publisher_zeromq/queue.py (hunk 3 above)
Comment (christian-marie) [inline ceilometer_publisher_zeromq/queue.py]: I can confirm that you should shutdown the original context at the minimum.

I also agree with @tranma in that we should really just zmq_disconnect and then zmq_connect again.

File: ceilometer_publisher_zeromq/queue.py (hunk 2 above)
Comment (fractalcat) [inline ceilometer_publisher_zeromq/queue.py]: Definitely does support name lookups for TCP URIs (at least ZMQ itself does, maybe the Python bindings don't or something?). It doesn't really matter.

File: ceilometer_publisher_zeromq/queue.py (hunk 3 above)
Comment (fractalcat) [inline ceilometer_publisher_zeromq/queue.py]: I wasn't sure if zmq_disconnect would reset the state of a REQ/REP socket pair. It probably does though.

File: ceilometer_publisher_zeromq/queue.py (hunk 2 above)
Comment (oswynb) [inline ceilometer_publisher_zeromq/queue.py]: http://api.zeromq.org/2-1:zmq-tcp

Ahh, so connects have DNS, binds do not. Shall update.

File: test_integration.py [hunk 4]
```diff
@@ -35,38 +35,23 @@ def stub_expected():
...
-    if message:
-        message.ack()
-        cleaned = json.loads(message.payload)
-        cleaned.pop('id', None)
-        test_queue.close()
-        assert cleaned == stub_expected()
+    new_pid = os.fork()
+    if new_pid == 0:
+        queue.QueuePublisher(
+            SplitResult(scheme = 'rabbitmq', netloc='', path='', query='', fragment='')
```
Comment (thsutton) [inline test_integration.py]: `rabbitmq`?

File: test_integration.py [hunk 5]
```diff
@@ -35,38 +35,23 @@ def stub_expected():
...
-        userid       = options.publisher_rabbit_user,
-        password     = options.publisher_rabbit_password,
-        virtual_host = options.publisher_rabbit_virtual_host,
-        port         = options.publisher_rabbit_port,
-    )
-    connection.connect()
+    context = zmq.Context()
+    socket  = context.socket(zmq.REP)
+    socket.bind("tcp://127.0.0.1:8282")
     #Use the publihser to publish a single message
```
Comment (thsutton) [inline test_integration.py]: speling

File: ceilometer_publisher_zeromq/queue.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        self.context.term()
+        self.connect()
+
+    def connect(self):
+        """connects to the collector"""
+        self.context = zmq.Context()
+        self.socket  = self.context.socket(zmq.REQ)
+        self.socket.connect("tcp://%s:%d" % (cfg.CONF.publisher_zeromq_host, cfg.CONF.publisher_zeromq_port))
+
+    def publish_sample(self, message):
```
Comment (thsutton) [inline ceilometer_publisher_zeromq/queue.py]: It seems odd to me that there's no exception handling? Does ceilometer handle that outside this class?

Comment (oswynb) [inline ceilometer_publisher_zeromq/queue.py]: With all the ZMQ use in publisher the only potentially failing bit is wrapped with the poll. Ceilometer does to continue after errors but it's basically, oh the publisher crashed, going to restart it.

Comment (thsutton) [inline ceilometer_publisher_zeromq/queue.py]: In that case, looks OK to me

Comment (fractalcat) [inline ceilometer_publisher_zeromq/queue.py]: @oswynb are you sure about that? I thought the publisher did not get restarted if it died with an uncaught exception.

Comment (oswynb) [inline ceilometer_publisher_zeromq/queue.py]: @fractalcat it seems that the agent catches it (or something) logs it, then continues.

2015-02-27 00:43:57.990 8775 ERROR ceilometer.pipeline [-] Pipeline meter_sink: Continue after error from publisher <ceilometer_publisher_zeromq.queue.QueuePublisher object at 0x2236a50>

**Pull request merged**